resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/FailureTracker.scala

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.deploy.yarn
+
+import scala.collection.mutable
+
+import org.apache.spark.SparkConf
+import org.apache.spark.internal.Logging
+import org.apache.spark.util.{Clock, SystemClock}
+
+/**
+ * FailureTracker is responsible for tracking executor failures both for each host separately
+ * and for all host altogether.
+*/
+private[spark] class FailureTracker(
+    sparkConf: SparkConf,
+    var clock: Clock = new SystemClock) extends Logging {
```

Comment:
But in that case we have to extend YarnAllocator constructor with a Clock too as in the YarnAllocatorSuite at the "window based failure executor counting" test method a manual clock is used. Of course that can be an optional parameter initialised to SystemClock by default.

Reply:
Yes, that would be nice. That's what `FsHistoryProvider` does for example (although it uses two constructors instead of a default value - either is fine really).